### DISCUSSION TITLE
Special Functions

### DISCUSSION THREAD

**Discussion by FRPProject** [OpenI6X/opentx#326] [Q&A] [answered]
I have repurposed the throttle-trim & rudder-trim buttons to become 3-position switches. To do so, I have used up all special function available lines. Is it possible to increase the number of the special functions in future release? Thanks.

**Comment by ajjjjjjjj** [comment, marked as answer]
Yes, glad you asked, i was thinking what to increase after recently recovered RAM and flash due to compiler change.
How many more is needed in your case?

**Reply by FRPProject** [reply]
Thanks for replying back. 

I'd say 10 more lines would be great, but the more the better; I was thinking that it would be nice to give the user the flexibility of determining how to use the memory space of the radio by choosing how many lines of "Special Functions" they may need. You see, I do have lots of memory left on my i6x as I configured all my quads/planes on one model on the OpenTX. I do most settings on Betaflight and INAV if possible, but repurposing the trim buttons is a great idea that I recently discovered and implemented.

**Reply by ajjjjjjjj** [reply]
It's not about eeprom memory.

Changing from 11 to 16, maybe 18 is possible.

**Reply by FRPProject** [reply]
Sounds good; Upgrading to 16 or 18 is much better than no line left in my case :)

**Reply by ajjjjjjjj** [reply]
You may test it, it will reset your eeprom due to changes: https://github.com/OpenI6X/opentx/pull/328

**Reply by FRPProject** [reply]
Thanks, I'll be back in a few :)

**Reply by FRPProject** [reply]
Works GREAT 💯  Thanks to you, I have now two solid multi-position switches on the throttle/rudder-trim buttons adjusting G1 & G2 written on CH11 & CH12. The buttons loop (from 100 to -100 and vice versus with a beep at each limits). On the radio startup, they are set to -100 so that I know where they are positioned. Now, there are 2 lines left out of 16 that I am going to use for Telemetry alarms (VBat, ...).  Once again, thanks for your time and generosity @ajjjjjjjj

**Reply by ajjjjjjjj** [reply]
You're welcome! merged.

**Reply by ajjjjjjjj** [reply]
You'll need to manually reset eeprom data on next release or current master because i have increased telemetry slots to 30.
It looks like you have some advanced configuration. Is it possible that you share your eeprom copy for analysis purposes on how RAM buffers behave?

**Reply by FRPProject** [reply]
Sure, absolutely; I am actually planning to make an eeprom backup today. I'll share it with you soon

**Reply by FRPProject** [reply]
@ajjjjjjjj Here is the backup, I have also included the custom i6x-OpenTX firmware that I compiled based on your modifications (https://github.com/OpenI6X/opentx/pull/328) (I am using backlight mod +Bluetooth telemetry with HC05 module)

[Backup_EEPROM.zip](https://github.com/OpenI6X/opentx/files/10300491/Backup_EEPROM.zip)

**Reply by FRPProject** [reply]
@ajjjjjjjj Let me know what you find about the RAM analysis. Thanks.

**Reply by ajjjjjjjj** [reply]
So i took a look and it looks like menu task RAM usage is not increasing even with relatively complex setup. Thanks.

Regarding your setup i like it a lot - What is Edge function guarding from? From accidental trim touch?

BTW. I increased special functions to 18 for next release, but remember that you'll have to reset eeprom manually.

**Reply by FRPProject** [reply]
Honestly, I was a bit worried about pushing the radio to its limit by using these special functions :)

I am primarily using the Edge function to set the value (associated with configured channels) to max instead of gradually increasing by "+50%". This allows the long press to be the max value (=100%) but the short press just as a way of increasing the value by 50%. 

This a good news. It seems I cannot get enough special functions out of OpenTX :) I already have some ideas on how to use the newly added special functions :))  Thanks a lot.
